### DISCUSSION TITLE
Clarification Requested on Terminology: Items, Cells, and Rows in SlickGrid

### DISCUSSION THREAD

**Discussion by pbower** [6pac/SlickGrid#951] [Q&A]
Hi there,

I've recently integrated SlickGrid into a project and found myself a bit confused by the terminology used in the DataView component, specifically around 'items', 'cells', and 'rows'. Initially, I assumed 'items' referred to 'cells', but after diving deeper, I realise this might not be the case. I'm hoping someone can help clarify these concepts for me.

From the [DataView documentation](https://github.com/6pac/SlickGrid/wiki/DataView), I understand that:

- Items might refer to the full rows of the underlying dataset.
- Rows seem to pertain to the rows within the DataView subset that's been rendered.

However, there are a few points in the documentation and codebase that I find conflicting or unclear:

- **Terminology of 'Item'**: The term 'item' suggests a singular entity, which led me to think of it as a 'cell'. The definition of an 'item' isn't explicitly stated in the DataView [Getting Started](https://github.com/6pac/SlickGrid/wiki/Getting-Started), and follows an introduction of 'Rows' referring to what I now believe 'Items' might truly refer to. Could you confirm if an 'item' indeed represents a full row in the underlying dataset?

- **Function getItems()**: The code comments describe this as "/** Get all DataView Items */". I therefore initially thought this referred to all cells in the visible grid. Is this actually retrieving the entire array of rows in the underlying dataset? If so, how does this work efficiently with very large datasets? 

- **Updating Cell Values**: For spreadsheet-like functionality where cell values need to be updated, what is the recommended approach if not using updateSingleItem (assuming this refers to rows, not cells ?)?

- **Clarification on 'id' vs. 'Index':** Is 'Index' referring to numerical row positions starting from 0, or does it relate to the 'id's assigned to each row (item) (like in pandas)?

Any insights into these areas would be greatly appreciated, especially regarding efficient cell updates and large dataset handling in a custom DataView implementation. 

Thank you so much for your time and assistance, I really appreciate it!

**Comment by ghiscoding** [comment]
`item` is your data object, sometime named `dataContext`, sometime `item`. The word `cell` is used for something else, most often as `row` index and `cell` which is actually the column index. Another confusing part is that sometime a cell refers to an object containing cell & column, e.g. `getActiveCell()` returns a cell object like `{ row: 2, cell: 4 }`

The thing you need to be aware in SlickGrid is that the DataView is optional and in all of my libs I chose to always use it. So going back to the fact that it's optional, SlickGrid can be used by itself and for that reason some methods had to provide different name to avoid conflicts. In SlickGrid we call `getData()` but in DataView we call `getItems()` which is the full array of data. The reason I chose to always use the DataView is because it provides more functionalities that SlickGrid doesn't have, the Grouping feature is the biggest reason. 

To update the data, with the DataView you use `updateItem()`, to update multiple of them you call `updateItems()` and if you have a lot of them you can also use the DataView transaction approach by using `beginUpdate()`, `updateItems()` then close it with `endUpdate()`. The transactional way is simply to render the UI once instead of rerendering the UI on every update which is more expensive on the browser paint process.

For the `id` vs `index`, it's most often the `id` that is used which can also be customized when instantiating the DataView with another id, for example `guid`, that would be `setItems([...], 'guid');`.

So I think the confusing part might be the fact that SlickGrid sometime have similar functions but they had to be named differently so I would guess it was hard to come up with different names representing the same kind of functionality. Also worth to know that some variable usage are not entirely consistent across the project, e.g. sometime we see `item`, sometime we see `dataContext`

**Reply by pbower** [reply]
Thanks, that's super helpful. 

I hope you won't mind one more quick question ? Is getItems() called often by the DataView? For context, I'm working over RPC mechanisms, and therefore want to avoid pulling too much data. I.e, my custom data view is simply RPC calls in many places (so far).
Given Dataview is 'Slickgrid's view of the world', still wrapping my head around some of these mechanics, given the library has been highly optimised for performance.
Is it fair to say that Dataview uses these method as a basis for getting access to the full data object that it then subsets through access methods to return specific ranges accordingly ?

I should be good from here, and appreciate the support.

**Reply by ghiscoding** [reply]
I'm not exactly sure what you're asking, `getItems()` is never used internally since it has a direct items reference which is what it uses internally. There was some optimization made in the past, take a look at this [Example - Optimizing Updates](https://6pac.github.io/SlickGrid/examples/example-optimizing-updates.html). The previous code was using a simple array but the new code uses [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) which is faster than an Array and also has the benefit of never having duplicate entries since that is the principle of Map. I've never used RPC and barely know anything about it, so can't comment

**Reply by pbower** [reply]
Awesome, thanks, that clarifies it.

**Reply by 6pac** [reply]
Just a comment - it might be worth looking at the AJAX examples. They use a remote data source, and the strategy is to use paging to fetch a cluster of rows at a time. For example, if the grid is viewing 14 rows, perhaps fetch the current page plus three pages backwards and forwards. The client side data source can keep track of what has and hasn't been fetched. This can get complicated when you throw optimistic/pessimistic concurrency and filtering and sorting into the mix.

**Reply by pbower** [reply]
Brilliant, will do. Cheers!
